File: CODE TOUR/Code Tour 2023 - Code Tour Đà Nẵng/D.cpp
#include <iostream>
#include <vector>
#include <limits>

using namespace std;

int find_max_value(vector<int> &a, int w1, int w2)
{
    int n = a.size();
    int max_sum = numeric_limits<int>::min();

    for (int i1 = 0; i1 < n - 4; ++i1)
    {
        for (int i2 = i1 + 1; i2 < n - 3; ++i2)
        {
            for (int i3 = i2 + 1; i3 < n - 2; ++i3)
            {
                for (int i4 = i3 + 1; i4 < n - 1; ++i4)
                {
                    for (int i5 = i4 + 1; i5 < n; ++i5)
                    {
                        int current_sum = w1 * (a[i1] + a[i5]) +
                                          w2 * (a[i2] + a[i4]) + a[i3];
                        if (current_sum > max_sum)
                        {
                            max_sum = current_sum;
                        }
                    }
                }
            }
        }
    }

    return max_sum;
}

int find_max_value(vector<int> &a, int w1)
{
    int n = a.size();
    int max_value = numeric_limits<int>::min();

    for (int i1 = 0; i1 < n - 2; ++i1)
    {
        for (int i2 = i1; i2 < n - 1; ++i2)
        {
            for (int i3 = i2 + 1; i3 < n; ++i3)
            {
                int current_value = (a[i1] + a[i3]) * w1 + a[i2];
                if (current_value > max_value)
                {
                    max_value = current_value;
                }
            }
        }
    }

    return max_value;
}

int main()
{
    int n, w1, w2;
    cin >> n >> w1 >> w2;

    vector<int> a(n);
    for (int i = 0; i < n; ++i)
    {
        cin >> a[i];
    }

    cout << find_max_value(a, w1, w2) << endl;

    return 0;
}

File: CODE TOUR/Code Tour 2023 - Code Tour Đà Nẵng/D.py
n, w1, w2 = list(map(int, input().split()))
a = list(map(int, input().split()))


def find_max_value(a, w1, w2):
    n = len(a)
    max_value = float('-inf')

    for i1 in range(n):
        for i2 in range(i1, n):
            for i3 in range(i2, n):
                for i4 in range(i3, n):
                    for i5 in range(i4, n):
                        current_value = (a[i1] + a[i5]) * \
                            w1 + (a[i2] + a[i4]) * w2 + a[i3]
                        if current_value > max_value:
                            max_value = current_value

    return max_value


if w1 == w2 == 0:
    print(find_max_value(a, w1, w2))
elif w1 == 0 and w2 < 0:
    res = -10**19
    for i in range(n):
        res = max(res, w2*2*a[i]+a[i])
    for i in range(n):
        for j in range(i, n):
            for k in range(j, n):
                res = max(res, w2*(a[i]+a[k])+a[j])
    print(res)
else:
    print(find_max_value(a, w1, w2))
